// @acala-network/eth-providers/lib/utils/gqlTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionReceiptsOrderBy = exports.SubqueriesOrderBy = exports.LogsOrderBy = void 0;
/** Methods to use when ordering `Log`. */
var LogsOrderBy;
(function (LogsOrderBy) {
    LogsOrderBy["AddressAsc"] = "ADDRESS_ASC";
    LogsOrderBy["AddressDesc"] = "ADDRESS_DESC";
    LogsOrderBy["BlockHashAsc"] = "BLOCK_HASH_ASC";
    LogsOrderBy["BlockHashDesc"] = "BLOCK_HASH_DESC";
    LogsOrderBy["BlockNumberAsc"] = "BLOCK_NUMBER_ASC";
    LogsOrderBy["BlockNumberDesc"] = "BLOCK_NUMBER_DESC";
    LogsOrderBy["CreatedAtAsc"] = "CREATED_AT_ASC";
    LogsOrderBy["CreatedAtDesc"] = "CREATED_AT_DESC";
    LogsOrderBy["DataAsc"] = "DATA_ASC";
    LogsOrderBy["DataDesc"] = "DATA_DESC";
    LogsOrderBy["IdAsc"] = "ID_ASC";
    LogsOrderBy["IdDesc"] = "ID_DESC";
    LogsOrderBy["LogIndexAsc"] = "LOG_INDEX_ASC";
    LogsOrderBy["LogIndexDesc"] = "LOG_INDEX_DESC";
    LogsOrderBy["Natural"] = "NATURAL";
    LogsOrderBy["PrimaryKeyAsc"] = "PRIMARY_KEY_ASC";
    LogsOrderBy["PrimaryKeyDesc"] = "PRIMARY_KEY_DESC";
    LogsOrderBy["ReceiptIdAsc"] = "RECEIPT_ID_ASC";
    LogsOrderBy["ReceiptIdDesc"] = "RECEIPT_ID_DESC";
    LogsOrderBy["RemovedAsc"] = "REMOVED_ASC";
    LogsOrderBy["RemovedDesc"] = "REMOVED_DESC";
    LogsOrderBy["TopicsAsc"] = "TOPICS_ASC";
    LogsOrderBy["TopicsDesc"] = "TOPICS_DESC";
    LogsOrderBy["TransactionHashAsc"] = "TRANSACTION_HASH_ASC";
    LogsOrderBy["TransactionHashDesc"] = "TRANSACTION_HASH_DESC";
    LogsOrderBy["TransactionIndexAsc"] = "TRANSACTION_INDEX_ASC";
    LogsOrderBy["TransactionIndexDesc"] = "TRANSACTION_INDEX_DESC";
    LogsOrderBy["UpdatedAtAsc"] = "UPDATED_AT_ASC";
    LogsOrderBy["UpdatedAtDesc"] = "UPDATED_AT_DESC";
})(LogsOrderBy = exports.LogsOrderBy || (exports.LogsOrderBy = {}));
/** Methods to use when ordering `Subquery`. */
var SubqueriesOrderBy;
(function (SubqueriesOrderBy) {
    SubqueriesOrderBy["CreatedAtAsc"] = "CREATED_AT_ASC";
    SubqueriesOrderBy["CreatedAtDesc"] = "CREATED_AT_DESC";
    SubqueriesOrderBy["DbSchemaAsc"] = "DB_SCHEMA_ASC";
    SubqueriesOrderBy["DbSchemaDesc"] = "DB_SCHEMA_DESC";
    SubqueriesOrderBy["HashAsc"] = "HASH_ASC";
    SubqueriesOrderBy["HashDesc"] = "HASH_DESC";
    SubqueriesOrderBy["IdAsc"] = "ID_ASC";
    SubqueriesOrderBy["IdDesc"] = "ID_DESC";
    SubqueriesOrderBy["NameAsc"] = "NAME_ASC";
    SubqueriesOrderBy["NameDesc"] = "NAME_DESC";
    SubqueriesOrderBy["Natural"] = "NATURAL";
    SubqueriesOrderBy["NetworkAsc"] = "NETWORK_ASC";
    SubqueriesOrderBy["NetworkDesc"] = "NETWORK_DESC";
    SubqueriesOrderBy["NetworkGenesisAsc"] = "NETWORK_GENESIS_ASC";
    SubqueriesOrderBy["NetworkGenesisDesc"] = "NETWORK_GENESIS_DESC";
    SubqueriesOrderBy["NextBlockHeightAsc"] = "NEXT_BLOCK_HEIGHT_ASC";
    SubqueriesOrderBy["NextBlockHeightDesc"] = "NEXT_BLOCK_HEIGHT_DESC";
    SubqueriesOrderBy["PrimaryKeyAsc"] = "PRIMARY_KEY_ASC";
    SubqueriesOrderBy["PrimaryKeyDesc"] = "PRIMARY_KEY_DESC";
    SubqueriesOrderBy["UpdatedAtAsc"] = "UPDATED_AT_ASC";
    SubqueriesOrderBy["UpdatedAtDesc"] = "UPDATED_AT_DESC";
    SubqueriesOrderBy["VersionAsc"] = "VERSION_ASC";
    SubqueriesOrderBy["VersionDesc"] = "VERSION_DESC";
})(SubqueriesOrderBy = exports.SubqueriesOrderBy || (exports.SubqueriesOrderBy = {}));
/** Methods to use when ordering `TransactionReceipt`. */
var TransactionReceiptsOrderBy;
(function (TransactionReceiptsOrderBy) {
    TransactionReceiptsOrderBy["BlockHashAsc"] = "BLOCK_HASH_ASC";
    TransactionReceiptsOrderBy["BlockHashDesc"] = "BLOCK_HASH_DESC";
    TransactionReceiptsOrderBy["BlockNumberAsc"] = "BLOCK_NUMBER_ASC";
    TransactionReceiptsOrderBy["BlockNumberDesc"] = "BLOCK_NUMBER_DESC";
    TransactionReceiptsOrderBy["ContractAddressAsc"] = "CONTRACT_ADDRESS_ASC";
    TransactionReceiptsOrderBy["ContractAddressDesc"] = "CONTRACT_ADDRESS_DESC";
    TransactionReceiptsOrderBy["CreatedAtAsc"] = "CREATED_AT_ASC";
    TransactionReceiptsOrderBy["CreatedAtDesc"] = "CREATED_AT_DESC";
    TransactionReceiptsOrderBy["CumulativeGasUsedAsc"] = "CUMULATIVE_GAS_USED_ASC";
    TransactionReceiptsOrderBy["CumulativeGasUsedDesc"] = "CUMULATIVE_GAS_USED_DESC";
    TransactionReceiptsOrderBy["FromAsc"] = "FROM_ASC";
    TransactionReceiptsOrderBy["FromDesc"] = "FROM_DESC";
    TransactionReceiptsOrderBy["GasUsedAsc"] = "GAS_USED_ASC";
    TransactionReceiptsOrderBy["GasUsedDesc"] = "GAS_USED_DESC";
    TransactionReceiptsOrderBy["IdAsc"] = "ID_ASC";
    TransactionReceiptsOrderBy["IdDesc"] = "ID_DESC";
    TransactionReceiptsOrderBy["LogsBloomAsc"] = "LOGS_BLOOM_ASC";
    TransactionReceiptsOrderBy["LogsBloomDesc"] = "LOGS_BLOOM_DESC";
    TransactionReceiptsOrderBy["Natural"] = "NATURAL";
    TransactionReceiptsOrderBy["PrimaryKeyAsc"] = "PRIMARY_KEY_ASC";
    TransactionReceiptsOrderBy["PrimaryKeyDesc"] = "PRIMARY_KEY_DESC";
    TransactionReceiptsOrderBy["StatusAsc"] = "STATUS_ASC";
    TransactionReceiptsOrderBy["StatusDesc"] = "STATUS_DESC";
    TransactionReceiptsOrderBy["ToAsc"] = "TO_ASC";
    TransactionReceiptsOrderBy["ToDesc"] = "TO_DESC";
    TransactionReceiptsOrderBy["TransactionHashAsc"] = "TRANSACTION_HASH_ASC";
    TransactionReceiptsOrderBy["TransactionHashDesc"] = "TRANSACTION_HASH_DESC";
    TransactionReceiptsOrderBy["TransactionIndexAsc"] = "TRANSACTION_INDEX_ASC";
    TransactionReceiptsOrderBy["TransactionIndexDesc"] = "TRANSACTION_INDEX_DESC";
    TransactionReceiptsOrderBy["TypeAsc"] = "TYPE_ASC";
    TransactionReceiptsOrderBy["TypeDesc"] = "TYPE_DESC";
    TransactionReceiptsOrderBy["UpdatedAtAsc"] = "UPDATED_AT_ASC";
    TransactionReceiptsOrderBy["UpdatedAtDesc"] = "UPDATED_AT_DESC";
})(TransactionReceiptsOrderBy = exports.TransactionReceiptsOrderBy || (exports.TransactionReceiptsOrderBy = {}));
