// @acala-network/api/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.options = exports.defaultOptions = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _ormlApiDerive = require("@open-web3/orml-api-derive");

var _apiDerive = require("@acala-network/api-derive");

var _types = require("@acala-network/types");

const _excluded = ["types", "rpc", "typesAlias", "typesBundle", "signedExtensions"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

const defaultOptions = {
  types: _types.types,
  rpc: _types.rpc
};
exports.defaultOptions = defaultOptions;

const options = function (_ref) {
  var _acalaTypesBundle$spe, _typesBundle$spec, _acalaTypesBundle$spe2, _typesBundle$spec2, _acalaTypesBundle$spe3, _typesBundle$spec3;

  if (_ref === void 0) {
    _ref = {};
  }

  let {
    types = {},
    rpc = {},
    typesAlias = {},
    typesBundle = {},
    signedExtensions
  } = _ref,
      otherOptions = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return _objectSpread({
    types: _objectSpread({}, types),
    rpc: _objectSpread(_objectSpread({}, _types.rpc), rpc),
    typesAlias: _objectSpread(_objectSpread({}, _types.typesAlias), typesAlias),
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    derives: _objectSpread(_objectSpread({}, _ormlApiDerive.derive), _apiDerive.derive),
    typesBundle: _objectSpread(_objectSpread({}, typesBundle), {}, {
      spec: _objectSpread(_objectSpread({}, typesBundle.spec), {}, {
        acala: _objectSpread(_objectSpread({}, _types.typesBundle === null || _types.typesBundle === void 0 ? void 0 : (_acalaTypesBundle$spe = _types.typesBundle.spec) === null || _acalaTypesBundle$spe === void 0 ? void 0 : _acalaTypesBundle$spe.acala), typesBundle === null || typesBundle === void 0 ? void 0 : (_typesBundle$spec = typesBundle.spec) === null || _typesBundle$spec === void 0 ? void 0 : _typesBundle$spec.acala),
        mandala: _objectSpread(_objectSpread({}, _types.typesBundle === null || _types.typesBundle === void 0 ? void 0 : (_acalaTypesBundle$spe2 = _types.typesBundle.spec) === null || _acalaTypesBundle$spe2 === void 0 ? void 0 : _acalaTypesBundle$spe2.mandala), typesBundle === null || typesBundle === void 0 ? void 0 : (_typesBundle$spec2 = typesBundle.spec) === null || _typesBundle$spec2 === void 0 ? void 0 : _typesBundle$spec2.mandala),
        karura: _objectSpread(_objectSpread({}, _types.typesBundle === null || _types.typesBundle === void 0 ? void 0 : (_acalaTypesBundle$spe3 = _types.typesBundle.spec) === null || _acalaTypesBundle$spe3 === void 0 ? void 0 : _acalaTypesBundle$spe3.karura), typesBundle === null || typesBundle === void 0 ? void 0 : (_typesBundle$spec3 = typesBundle.spec) === null || _typesBundle$spec3 === void 0 ? void 0 : _typesBundle$spec3.mandala)
      })
    }),
    signedExtensions: _objectSpread(_objectSpread({}, _types.signedExtensions), signedExtensions)
  }, otherOptions);
};

exports.options = options;
